planner/core/rule_join_reorder.go

```diff
@@ -0,0 +1,168 @@
+)
+
+// extractJoinGroup extracts all the join nodes connected with continuous
+// InnerJoins to construct a join group. This join group is further used to
+// construct a new join order based on a greedy algorithm.
+//
+// For example: "InnerJoin(InnerJoin(a, b), LeftJoin(c, d))"
+// results in a join group {a, b, LeftJoin(c, d)}.
+func extractJoinGroup(p LogicalPlan) (group []LogicalPlan, eqEdges []*expression.ScalarFunction, otherConds []expression.Expression) {
+	join, isJoin := p.(*LogicalJoin)
+	if !isJoin || join.preferJoinType > uint(0) || join.JoinType != InnerJoin || join.StraightJoin {
+		return []LogicalPlan{p}, nil, nil
+	}
+
+	lhsGroup, lhsEqualConds, lhsOtherConds := extractJoinGroup(join.children[0])
+	rhsGroup, rhsEqualConds, rhsOtherConds := extractJoinGroup(join.children[1])
+
+	group = append(group, lhsGroup...)
+	group = append(group, rhsGroup...)
+	eqEdges = append(eqEdges, join.EqualConditions...)
+	eqEdges = append(eqEdges, lhsEqualConds...)
+	eqEdges = append(eqEdges, rhsEqualConds...)
+	otherConds = append(otherConds, join.OtherConditions...)
+	otherConds = append(otherConds, lhsOtherConds...)
+	otherConds = append(otherConds, rhsOtherConds...)
+	return group, eqEdges, otherConds
+}
+
+type joinReOrderSolver struct {
+}
+
+type jrNode struct {
+	p       LogicalPlan
+	cumCost float64
+}
+
+func (s *joinReOrderSolver) optimize(p LogicalPlan) (LogicalPlan, error) {
+	return s.optimizeRecursive(p.context(), p)
+}
+
+// optimizeRecursive first checks that whether the current plan tree is start with a join group.
+// If so it will call the reorder algorithm to reorder the join group.
+func (s *joinReOrderSolver) optimizeRecursive(ctx sessionctx.Context, p LogicalPlan) (LogicalPlan, error) {
+	var err error
+	curJoinGroup, eqEdges, otherConds := extractJoinGroup(p)
+	if len(curJoinGroup) > 1 {
+		for i := range curJoinGroup {
+			curJoinGroup[i], err = s.optimizeRecursive(ctx, curJoinGroup[i])
+			if err != nil {
+				return nil, err
+			}
+		}
+		baseGroupSolver := &baseSingleGroupJoinOrderSolver{
+			ctx:        ctx,
+			otherConds: otherConds,
+		}
+		groupSolver := &joinReorderGreedySingleGroupSolver{
+			baseSingleGroupJoinOrderSolver: baseGroupSolver,
+			eqEdges:                        eqEdges,
+		}
+		p, err = groupSolver.solve(curJoinGroup)
+		if err != nil {
+			return nil, err
+		}
+		return p, nil
+	}
+	newChildren := make([]LogicalPlan, 0, len(p.Children()))
+	for _, child := range p.Children() {
+		newChild, err := s.optimizeRecursive(ctx, child)
+		if err != nil {
+			return nil, err
+		}
+		newChildren = append(newChildren, newChild)
+	}
+	p.SetChildren(newChildren...)
+	return p, nil
+}
+
+type baseSingleGroupJoinOrderSolver struct {
+	ctx          sessionctx.Context
+	curJoinGroup []*jrNode
+	otherConds   []expression.Expression
+}
+
+// baseNodeCumCost calculate the cumulative cost of the node in the join group.
+func (s *baseSingleGroupJoinOrderSolver) baseNodeCumCost(groupNode LogicalPlan) float64 {
```

Comment:
There is already a `base` keyword in `baseSingleGroupJoinOrderSolver`, how about renaming `baseNodeCumCost` to `calcNodeCumCost`?

And I found `baseSingleGroupJoinOrderSolver.calcJoinCumCost()` can be removed. In `joinReorderGreedySingleGroupSolver.constructConnectedJoinTree()`, `newJoin` is constructed with `curJoinTree` and `node`, so we can directly call `baseSingleGroupJoinOrderSolver.baseNodeCumCost(newJoin)` to achieve the same purpose.

Reply:
When join tree is large, use `baseNodeCumCost` to search the subtree again and again is not a good idea...